Graphql integration

ikagotso opened the pull request.

ikagotso committed:
GraphQL integration initial commit, added schema and controllers

ikagotso committed:
Merge remote-tracking branch 'origin/main' into graphql_integration

# Conflicts:
#	build.gradle

ikagotso committed:
modified GraphQL controllers according to mapstruct, added AbstractGraphQLController,
added unit tests for GraphQL

ikagotso committed:
fixed formatting

ikagotso committed:
added CommonGraphQLControllerTest for custom paging testing, added helper methods in TestUtils

ikagotso committed:
fixed code smell for AbstractGraphQLTest, modified graphql input data mapping using mapstruct

ikagotso committed:
Merge remote-tracking branch 'origin/main' into graphql_integration

ikagotso committed:
minor warning fix

ikagotso committed:
added description

ikagotso committed:
added test for complete coverage

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

nkvaratskhelia committed:
added graphql schema printing
- added spring.graphql.schema.printer.enabled prop in application.properties
- added Api subsection under the Web section in project-description.adoc
- removed GraphQL subsection under the Data section in project-description.adoc
- added graphql props in the Configuration > Properties subsection in project-setup.adoc
- reordered spring-boot-starter-graphql dependency

nkvaratskhelia committed:
added GraphQL to Web section in project

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

# Conflicts:
#	build.gradle

nkvaratskhelia commented on build.gradle:
```diff
@@ -114,7 +116,9 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
+    implementation 'org.springframework.boot:spring-boot-starter-graphql'
     implementation 'org.springframework.boot:spring-boot-starter-web'
+    compileOnly "com.graphql-java:graphql-java-extended-scalars:${graphqlJavaExtendedScalarsVersion}"
```
First of all, go to your application `Run/Debug Configurations` and remove the `Add dependencies with "provided" scope to classpath` option:

<img width="843" alt="image" src="https://user-images.githubusercontent.com/16168235/178914837-055c37ec-e929-416e-833d-e36f910b4edb.png">

After this, if you run the app, it will fail to start because of the `java.lang.ClassNotFoundException: graphql.scalars.ExtendedScalars` error. So you need to change the scope of the `graphql-java-extended-scalars` dependency to `implementation`. `compileOnly` worked in your IDE, but would have failed in production, because the `Add dependencies with "provided" scope to classpath` option only applies in the context of your IDE. CI tests didn't fail because you have this depedency additionally declared as `testRuntimeOnly`.

Action points:
* Change the scope of the dependency to implementation
* Remove the testRuntimeOnly scope of the dependency

ikagotso replied:
Done.

ikagotso committed:
removed spring-graphql-test dependency version, changed graphql-java-extended-scalars dependency type to implementation.

nkvaratskhelia committed:
reordered versions & dependencies

ikagotso commented on build.gradle:
```diff
@@ -101,6 +101,8 @@ ext {
     set('springBootAdminVersion', '2.7.2')
     set('springdocOpenapiVersion', '1.6.9')
+    set('graphqlJavaExtendedScalarsVersion', '18.1')
+    set('springGraphqlTestVersion', '1.0.0')
```
Done. Thank you for complete info.

nkvaratskhelia committed:
formatting

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

# Conflicts:
#	build.gradle

nkvaratskhelia committed:
formatting

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

nkvaratskhelia committed:
formatting

nkvaratskhelia committed:
formatting

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

# Conflicts:
#	src/main/java/com/example/universe/simulator/entityservice/mappers/GalaxyMapper.java
#	src/main/java/com/example/universe/simulator/entityservice/mappers/MoonMapper.java
#	src/main/java/com/example/universe/simulator/entityservice/mappers/PlanetMapper.java
#	src/main/java/com/example/universe/simulator/entityservice/mappers/StarMapper.java

nkvaratskhelia committed:
switched to inputs from the inputs package in graphql controllers
- removed validator field in graphql controllers
- added javax validation for inputs in add/update methods in graphql controllers
- removed custom validation in add/update methods in graphql controllers
- removed input records in graphql controllers
- removed validator field in graphql controller tests
- replaced dto param with input in buildInputMapForXXXAdd/Update methods in TestUtils

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

nkvaratskhelia committed:
aligned graphql & rest controllers
- fixed logging in graphql controllers
- changed return type of assemblePageRequest method to Pageable in AbstractGraphQLController

nkvaratskhelia commented on src/main/resources/graphql/schema.graphqls:
```diff
@@ -0,0 +1,122 @@
+}
+
+type GalaxyDto {
+    id: ID!
```
Why are id fields marked as required in Dtos?

ikagotso committed:
Removed AbstractGraphQLTest, added new mapping for PageInput using mapstruct, modified graphql schema to use defaults.

nkvaratskhelia committed:
added validation to page input
- added javax validation in PageInput
- reworked sort mapping in toPageable method in PageInputMapper

nkvaratskhelia committed:
formatting

nkvaratskhelia committed:
fixed comment in CommonGraphQLControllerTest

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

# Conflicts:
#	src/test/java/com/example/universe/simulator/entityservice/common/utils/TestUtils.java

nkvaratskhelia committed:
renamed mapper fields in graphql controller tests

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

nkvaratskhelia committed:
added graphql schema endpoint in web > api section in project-description.adoc

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

nkvaratskhelia committed:
reworked common graphql controller tests
- reworked tests in CommonGraphQLControllerTest
- removed getSpaceEntityPageableWithoutSorting, buildGalaxyWithName & buildInputMapForOnlyPaging methods in TestUtils
- added buildDefaultPageInput method in TestUtils

ikagotso committed:
moved GraphQLConfig import to AbstractGraphQLTest

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

nkvaratskhelia committed:
added static modifier from buildInputMapForXXX methods in TestUtils

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

nkvaratskhelia commented on src/main/resources/graphql/schema.graphqls:
```diff
@@ -0,0 +1,122 @@
+    direction: Direction = ASC
+}
+
+enum Direction {
```
This enum should be renamed to `SortDirection`, as we might have other direction types.

ikagotso replied:
It's named "Direction" intentionally, to comply with Sort.Direction. Otherwise new enum should be created on java side.

nkvaratskhelia replied:
No need for new java enum. Should work as is after you rename it to `SortDirection`.

nkvaratskhelia committed:
added common graphql controller tests
- added defaultSortDirection test in CommonGraphQLControllerTest
- reworked existing methods in CommonGraphQLControllerTest
- added @Valid to pageInput parameter in get list methods in graphql controllers
- added @Valid to SortOrder in PageInput
- changed public modifier to private for DEFAULT_PAGE_NUMBER & DEFAULT_PAGE_SIZE constants in TestUtils
- removed DEFAULT_SORT constant in TestUtils
- reworked paging methods in TestUtils

ikagotso committed:
Renamed Direction enum to SortDirection in schema.graphqls

nkvaratskhelia committed:
renamed fields in CommonGraphQLControllerTest
- renamed service field to galaxyService in CommonGraphQLControllerTest
- renamed specificationBuilder field to galaxySpecificationBuilder in CommonGraphQLControllerTest

nkvaratskhelia committed:
moved pageInputMapper field from AbstractGraphQLTest to CommonGraphQLControllerTest

nkvaratskhelia committed:
Merge branch 'main' into graphql_integration

nkvaratskhelia committed:
renamed mapper field in graphql controllers

nkvaratskhelia committed:
reworked graphql controller tests
- added toDtoList method in mappers
- reworked buildInputMapForXXX methods in TestUtils
- added missing version field to buildStar, buildPlanet & buildMoon methods in TestUtils
- extracted pageInputMap variable in CommonGraphQLControllerTest methods
